### PR TITLE
`ledger-api-bench-tool` - item rate SLO [DPP-790]

### PR DESCRIPTION
**Pull request by kamil-da**
### Changes
- a new service-level objective for transactions, transaction trees and active contract streams - minimum item rate (`min-rate`)

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/metrics/objectives/ServiceLevelObjective.scala [hunk 1]
```diff
@@ -3,8 +3,34 @@
...
+  DelayMetric,
+  MetricValue,
+}
 
-trait ServiceLevelObjective[MetricValueType <: MetricValue] {
+sealed trait ServiceLevelObjective[MetricValueType <: MetricValue] {
   def isViolatedBy(metricValue: MetricValueType): Boolean
 }
+
+final case class MaxDelay(maxDelaySeconds: Long) extends ServiceLevelObjective[DelayMetric.Value] {
```
Comment (kamil-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/metrics/objectives/ServiceLevelObjective.scala:17]: That's only a refactor.

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Benchmark.scala [hunk 2]
```diff
@@ -28,7 +27,10 @@ object Benchmark {
       reportingPeriod: FiniteDuration,
       apiServices: LedgerApiServices,
       metricsReporter: MetricsReporter,
-  )(implicit ec: ExecutionContext, resourceContext: ResourceContext): Future[Unit] = {
+  )(implicit
+      ec: ExecutionContext,
+      resourceContext: ResourceContext,
+  ): Future[Either[String, Unit]] = {
```
Comment (kamil-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Benchmark.scala:33]: Changed this to differentiate between benchmark failure (like too slow stream) and an application failure.

**Review by nmarton-da — APPROVED** [review]
LGTM

Random note, not belongs to this PR:
Never looked this code-base closely, now I realized that we use typed actors as the metrics infrastructure: although it is a valid approach I am not sure this use case warrants the additional complexity.
Opinionated note: I would first use Futures, simple async vehicles, akka-streams, and would utilize akka bare actors (or frameworks respectively) as the last resort due to it's significant code, and conceptual overhead. (especially that you also had to build an integration zone to Future-s to integrate with the rest of the code)
Code hygiene note: We are not really using typed akka, not even akka for that matter in the sdk code-base, not sure this use case warrants the leap for bare actors. Was this vetted cross team?

**Comment by kamil-da** [comment]
I didn't mention the intent to use typed Actors in any other place than in the initial PR itself - sorry for that!
The motivation for using actors was that for me it seems a cleaner approach if I have to deal with state reads/updates from a gRPC stream and periodic tasks at the same time. But that's probably a matter of preference.

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Benchmark.scala (hunk 2 above)
Comment (kamil-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Benchmark.scala:33]: The differentiation is introduced not to log errors in the same form as exceptions (with the stack trace, etc.). I could introduce a specific exception class for errors and pattern match but the approach I proposed seems more like "FP".
@nmarton-da would you like me to change this in a follow-up PR?

**Pull request merged**